x/valset-pref/validator_set_test.go

```diff
@@ -147,6 +145,244 @@ func (s *KeeperTestSuite) TestIsValidatorSetEqual() {
+					DelegatedAmt:  defaultDelegationAmt,
+					UndelegateAmt: defaultDelegationAmt,
+					VRatio:        sdk.NewDec(1),
+				},
+			},
+		},
+		{
+			name:             "single validator, undelegate partial amount",
+			useSingleValPref: true,
+			undelegateAmt:    sdk.NewInt(50),
+			expectedValRatios: []valPref.ValRatio{
+				{
+					Weight:        sdk.NewDec(1),
+					DelegatedAmt:  defaultDelegationAmt,
+					UndelegateAmt: defaultDelegationAmt.Quo(sdk.NewInt(2)),
+					// 0.5 since we are undelegating half amount
+					VRatio: sdk.NewDecWithPrec(5, 1),
+				},
+			},
+		},
+		{
+			name:          "multiple validator, undelegate full amount",
+			undelegateAmt: defaultDelegationAmt,
+			expectedValRatios: []valPref.ValRatio{
+				{
+					Weight:        sdk.MustNewDecFromStr("0.333333333333333333"),
+					DelegatedAmt:  defaultDelegationAmt,
+					UndelegateAmt: sdk.NewInt(33),
+					VRatio:        sdk.MustNewDecFromStr("0.33"),
+				},
+				{
+					Weight:        sdk.MustNewDecFromStr("0.666666666666666667"),
+					DelegatedAmt:  defaultDelegationAmt,
+					UndelegateAmt: sdk.NewInt(66),
+					VRatio:        sdk.MustNewDecFromStr("0.66"),
+				},
+			},
+		},
+		{
+			name:          "multiple validator, undelegate partial amount",
+			undelegateAmt: defaultDelegationAmt.Quo(sdk.NewInt(2)),
+			expectedValRatios: []valPref.ValRatio{
+				{
+					Weight:       sdk.MustNewDecFromStr("0.333333333333333333"),
+					DelegatedAmt: defaultDelegationAmt,
+					// 1/3 of undelegating amount(50)
+					UndelegateAmt: sdk.NewInt(16),
+					VRatio:        sdk.MustNewDecFromStr("0.16"),
+				},
+				{
+					Weight:       sdk.MustNewDecFromStr("0.666666666666666667"),
+					DelegatedAmt: defaultDelegationAmt,
+					// 2/3 of undelegating amount(50)
+					UndelegateAmt: sdk.NewInt(33),
+					VRatio:        sdk.MustNewDecFromStr("0.33"),
+				},
+			},
+		},
+		{
+			name:             "error: not delegated",
+			undelegateAmt:    defaultDelegationAmt,
+			useSingleValPref: true,
+			notDelegated:     true,
+			expectedError:    true,
+		},
+	}
+	for _, test := range tests {
+		s.Run(test.name, func() {
+			s.SetupTest()
+			valAddrs := s.SetupMultipleValidators(3)
+			defaultDelegator := s.TestAccs[0]
+
+			var valsetPrefs []types.ValidatorPreference
+			if test.useSingleValPref {
+				valsetPrefs = []types.ValidatorPreference{
+					{
+						ValOperAddress: valAddrs[0],
+						Weight:         sdk.OneDec(),
+					},
+				}
+			} else { // other cases, we assume we are using val set pref with mutiple validators
+				valsetPrefs = []types.ValidatorPreference{
+					{
+						ValOperAddress: valAddrs[0],
+						Weight:         sdk.MustNewDecFromStr("0.333333333333333333"),
+					},
+					{
+						ValOperAddress: valAddrs[1],
+						Weight:         sdk.MustNewDecFromStr("0.666666666666666667"),
+					},
+				}
+			}
+
+			// set up delegation for each of the valset prefs
+			expectedTotalDelegatedAmt := sdk.ZeroDec()
+			if !test.notDelegated {
```

Comment:
double negative hurts my brain lol

Reply:
Haha, If we use negative fields we can avoid having field set to true for all the other cases